extensions/amp-anim/0.1/amp-anim.js

```diff
@@ -93,37 +94,32 @@ export class AmpAnim extends AMP.BaseElement {
       /* opt_removeMissingAttrs */ true
     );
     guaranteeSrcForSrcsetUnsupportedBrowsers(img);
-    return this.loadPromise(img);
+    return this.loadPromise(img).then(() => {
+      observeWithSharedInOb(this.element, (inViewport) =>
+        this.viewportCallback_(inViewport)
+      );
+    });
   }
 
   /** @override */
   firstLayoutCompleted() {
     // Keep the placeholder: amp-anim is using it to start/stop playing.
-    this.hasLoaded_ = true;
-    this.updateInViewport_();
-  }
-
-  /** @override */
-  viewportCallback(unusedInViewport) {
-    if (!this.hasLoaded_) {
-      // do nothing if element has not laid out.
-      return;
-    }
-    this.updateInViewport_();
   }
 
   /** @override */
   unlayoutCallback() {
+    unobserveWithSharedInOb(this.element);
```

Comment:
There's one more nuance I completely missed. In the old world, before `unlayoutCallback` is executed, the resources system would also call `viewportCallback(false)`. See [this code](https://github.com/ampproject/amphtml/blob/90a250baf86d2172ff8d819a517fa157bab385b2/src/service/resource.js#L1066). I don't think it matters most of the time, but it's important here. When it matters, I think this can simply be done in the `unlayoutCallback` like this:

```
unlayoutCallback() {
  unobserveWithSharedInOb(this.element);
  this.viewportCallback_(false);
}
```

Please check other cases.

Reply:
Nice catch! I didn't know that happened. Indeed this case is to hide the animating image so it stops taking up CPU/GPU?
I've updated this case and checked the other 7, only one of which seemed like it could be mildly important (3d-gltf).

Reply:
Ah, I just found that `pauseCallback()` should also trigger viewportCallback(false)

Reply:
Hm, the only one that also has a `pauseCallback` is gltf, and it doesn't seem super necessary in this case.